torch/csrc/jit/codegen/cuda/arith.h

```diff
@@ -18,51 +20,251 @@ namespace torch {
 namespace jit {
 namespace fuser {
 
+template <typename Head, typename... Tail>
+struct IsValidArithOpType {
+  static constexpr bool value = !std::is_base_of<Val, Head>::value
+      ? false
+      : IsValidArithOpType<Tail...>::value;
+};
+
+template <typename Type>
+struct IsValidArithOpType<Type> {
+  static constexpr bool value = std::is_base_of<Val, Type>::value;
+};
+
+template <typename Head, typename... Tail>
+struct HasTensorView {
+  static constexpr bool value = std::is_base_of<TensorView, Head>::value
+      ? true
+      : HasTensorView<Tail...>::value;
+};
+
+template <typename Type>
+struct HasTensorView<Type> {
+  static constexpr bool value = std::is_base_of<TensorView, Type>::value;
+};
+
+// Define return type of arithmetic operations. Currently, TensorView
+// when any of operands is TensorView, and Val otherwise.
+template <typename... OpTypes>
+struct ArithOpRetType {
+  using Type = typename std::
+      conditional<HasTensorView<OpTypes...>::value, TensorView, Val>::type;
+};
+
 // Promotion logic between two values, returns a new val from resulting type
 // promotion.
 TORCH_CUDA_API Val* promoteNew(Val* v1, Val* v2);
 
 // Insertion of casting op to dtype, returns new resulting val
 TORCH_CUDA_API Val* castOp(DataType dtype, Val* v1);
+TORCH_CUDA_API TensorView* castOp(DataType dtype, TensorView* v1);
 
 // Perform unary op type and return the output
 TORCH_CUDA_API Val* unaryOp(UnaryOpType type, Val* v1);
+TORCH_CUDA_API TensorView* unaryOp(DataType dtype, TensorView* v1);
 
 // Perform binary op type on v1 and v2 and return a type promoted output.
 // Mod, CeilDiv, and LT are considered Int only output operations for now.
 TORCH_CUDA_API Val* binaryOp(BinaryOpType type, Val* v1, Val* v2);
 
+// Overload for the case when operands are not Val*. Template matching
+// fails when their classes are not valid types as determined by
+// IsValidArithOpType
+template <
+    typename OpType1,
+    typename OpType2,
+    std::enable_if_t<IsValidArithOpType<OpType1, OpType2>::value>* = nullptr>
+TORCH_CUDA_API typename ArithOpRetType<OpType1, OpType2>::Type* binaryOp(
```

Comment:
Fancy code!

NITPICK: Does it make sense to return auto just for shorter code? since the logic of return type is explicitly written inside the static_cast for return.